nntrainer/compiler/ini_interpreter.cpp

```diff
@@ -0,0 +1,148 @@
+ * @author Parichay Kapoor <pk.kapoor@samsung.com>
+ * @author Jihoon Lee <jhoon.it.lee@samsung.com>
+ * @bug    No known bugs except for NYI items
+ */
+#include <ini_interpreter.h>
+
+#include <vector>
+
+#include <nntrainer_error.h>
+#include <nntrainer_log.h>
+#include <parse_util.h>
+#include <util_func.h>
+
+static constexpr const char *FUNC_TAG = "[IniInterpreter] ";
+
+static constexpr const char *UNKNOWN_STR = "UNKNOWN";
+static constexpr const char *NONE_STR = "NONE";
+static constexpr const char *MODEL_STR = "model";
+static constexpr const char *DATASET_STR = "dataset";
+static constexpr const char *OPTIMIZER_STR = "optimizer";
+
+namespace nntrainer {
+
+namespace {
+
+/** @todo:
+ *  1. deprecate tag dispatching along with #1072
+ *  2. deprecate getMergeableGraph (extendGraph should accept graph itself)
+ */
+class PlainLayer {};    /**< Plain Layer tag */
+class BackboneLayer {}; /**< Backbone Layer tag */
+
+template <typename T>
+static std::shared_ptr<Layer> section2layer(dictionary *ini,
+                                            const std::string &sec_name) {
+  /// NYI!
+  return nullptr;
+}
+
+static std::shared_ptr<GraphRepresentation>
+section2graph(dictionary *ini, const std::string &sec_name) {
+  /// NYI!
+  return nullptr;
+}
+
+bool graphSupported(const std::string &backbone_name) {
+  /// NYI!
+  return true;
+}
+
+std::vector<std::shared_ptr<Layer>>
+getMergeableGraph(std::shared_ptr<const GraphRepresentation>) {
+  return {};
+};
+
+} // namespace
+
+void IniGraphInterpreter::serialize(
+  std::shared_ptr<const GraphRepresentation> representation,
+  const std::string &out) {}
+
+std::shared_ptr<GraphRepresentation>
+IniGraphInterpreter::deserialize(const std::string &in) {
+
+  NNTR_THROW_IF(in.empty(), std::invalid_argument)
+    << FUNC_TAG << "given in file is empty";
+
+  NNTR_THROW_IF(!isFileExist(in), std::invalid_argument)
+    << FUNC_TAG << "given ini file does not exist";
+
+  dictionary *ini = iniparser_load(in.c_str());
+  NNTR_THROW_IF(!ini, std::runtime_error) << "loading ini failed";
+
+  auto freedict = [ini] { iniparser_freedict(ini); };
+
+  /** Get number of sections in the file */
+  int num_ini_sec = iniparser_getnsec(ini);
+  NNTR_THROW_IF_CLEANUP(num_ini_sec < 0, std::invalid_argument, freedict)
+    << FUNC_TAG << "invalid number of sections.";
+
+  std::shared_ptr<GraphRepresentation> graph =
+    std::make_shared<GraphRepresentation>();
+
+  try {
+    for (int idx = 0; idx < num_ini_sec; ++idx) {
+      auto sec_name_ = iniparser_getsecname(ini, idx);
+      NNTR_THROW_IF_CLEANUP(!sec_name_, std::runtime_error, freedict)
+        << "parsing a section name returned error, filename: " << in
+        << "idx: " << idx;
+
+      ml_logd("probing section_name: %s", sec_name_);
+      std::string sec_name(sec_name_);
+
+      if (istrequal(sec_name, MODEL_STR) || istrequal(sec_name, DATASET_STR) ||
+          istrequal(sec_name, OPTIMIZER_STR)) {
+        /// dedicated sections so skip
```

Comment:
where are these dedicated sections called?

Reply:
serializers are meant to work only on graph. Those are not related to graph. So they are not commonly treated inside an interpreter. unless we start to put those objects inside the graph.

Reply:
so, deserialize will be called in combination with `model loader` or `deserialize` will use model loader internally.
also, section2layer and section2graph will also call `model loader` internally?

Reply:
from #1079, it seems like `section2layer` does not call model loader, but parses the ini itself. So, I am little confused how is the work distributed - layers and graph are parsed by ini_interpreter and these sections by the `model loader`?
If my understanding is correct, it would be best to move all to ini_interpreter or let all be model loader and call it from here.

Reply:
1. `model loader`  delegates parsing graph to the interpreter by calling `interpreter->deserialize`
2. so graph parsing related logics are moving to `ini_interpreter`
3. `model loader` uses `interpreter` but `interpreter` won't use `model loader`
+. Training related objects will be still parsed inside the `model loader`

Reply:
The boundary of `interpreter` interface is parsing graph only. In this way, we can parse other type of backbones into nntrainer graph as well.

Reply:
Understood. Please add the dsecription above to `model loader` or `ini_interpreter`?

Reply:
Added to the `interpreter.h`